specify deny-warning types

mmou opened the pull request:
Dependent on modifications in rustsec (https://github.com/RustSec/rustsec-crate/pull/156)

========

This is an initial attempt at #195 that I'm hoping for some feedback on.

I erred towards not modifying the `rustsec` crate in the approach here because I wasn't sure about the crates that depend on `rustsec`, but I think it could be cleaner to do some work there, such as:
- have the `Report` struct have something like a `WarningInfo` struct (to replace the `Vec<Warning>`)
- have the `WarningInfo` struct essentially be a hashmap of `warning::Type` to `Vec<Warning>`
- modify `Warning` and/or `warning::Type` to better support the hashmap of warning types to lists of warnings; e.g. maybe ideally `warning::Type` enum (or something like it) would not hold data

mmou commented on src/config.rs:
```diff
@@ -128,6 +129,47 @@ impl OutputConfig {
 
+/// Warning kinds
+#[derive(Copy, Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub enum WarningKind {
```
here's an attempt at options for deny-warning

mmou commented on src/presenter.rs:
```diff
@@ -81,21 +81,63 @@ impl Presenter {
             self.print_vulnerability(vulnerability, &tree);
         }
 
+        let deny_unmaintained = self
```
Parsing `report.warnings` here instead of storing warnings by type in `Report` struct in `rustsec` crate

mmou commented on src/presenter.rs:
```diff
@@ -81,21 +81,63 @@ impl Presenter {
+        //  could instead do something like this (if Warning implemented Copy):
+        //
+        //  let (denied, others): (Vec<Warning>, Vec<Warning>) =
+        //     report.warnings.iter().partition(|&w| match w.kind {
```
Couldn't use `partition` because `Warning` doesn't implement `Copy`...

mmou commented on src/presenter.rs:
```diff
@@ -81,21 +81,63 @@ impl Presenter {
 
-            if self.config.deny_warnings {
-                status_err!("{} {} found", report.warnings.len(), warning_word);
+            status_err!("{} denied {} found", num_denied, warning_word);
```
Some decisions to be made about how to display this...

tarcieri replied:
Generally I've looked to `rustc` and `clippy` for ideas and conventions for this sort of thing.

They promote `warning:` -> `error:` in this case, and add a `note:` saying that deny warnings has been enabled.

mmou commented on src/presenter.rs:
```diff
@@ -106,7 +148,7 @@ impl Presenter {
             let msg = "this copy of cargo-audit has known advisories!";
 
-            if self.config.deny_warnings {
+            if self.config.deny_warnings.contains(&WarningKind::Other) {
```
Currently considering self advisories to be `WarningKind::Other`

tarcieri commented:
Thanks for taking a look at this!

I would suggest modifying the `rustsec` crate, and not worry too much about the downstream consumers, particularly for warnings. The only other notable crate that depends on `rustsec` directly is `cargo-deny`, and I don't think it consumes this information. Regardless, the API is still unstable so it's ok to change.

> have the WarningInfo struct essentially be a hashmap of warning::Type to Vec<Warning>

I think this would be the best approach.

mmou commented on Cargo.toml:
```diff
@@ -19,7 +19,7 @@ abscissa_core = "0.5.2"
 home = "0.5"
 lazy_static = "1"
-rustsec = { version = "0.18", features = ["dependency-tree"] }
+rustsec = { git = "https://github.com/mmou/rustsec-crate", branch = "mmou/issue-195-warnings", features = ["dependency-tree"] }
```
temporary

mmou replied:
waiting for next rustsec release.

mmou commented on src/commands/audit.rs:
```diff
@@ -213,7 +225,7 @@ impl Runnable for AuditCommand {
         }
 
-        let lockfile_path = self.file.as_ref().map(PathBuf::as_path);
+        let lockfile_path = self.file.as_deref();
```
clippy told me to use as_deref?

tarcieri commented:
Note: #221 includes changes which conflict with / duplicate some of this PR.

However, I think it should make it easier to implement, since it bumps the `rustsec` crate to the latest release.

mmou commented:
> Note: #221 includes changes which conflict with / duplicate some of this PR.
> 
> However, I think it should make it easier to implement, since it bumps the `rustsec` crate to the latest release.

Thanks for the FYI! I updated the branch, rebased off of master. 

I need your help fixing the Self-Audit CI step ` cargo run -- audit --deny-warnings --ignore RUSTSEC-2019-0031`; should be updated to: ` cargo run -- audit --deny-warnings all --ignore RUSTSEC-2019-0031`.
